src/views/CHQueryEditor.tsx

```diff
@@ -1,95 +1,91 @@
 
-  switch (query.queryType) {
-    case QueryType.SQL:
-      return (
-        <div data-testid="query-editor-section-sql">
-          <SQLEditor {...props} />
-        </div>
-      );
-    case QueryType.Builder:
-    default:
-      let newQuery: CHBuilderQuery = { ...query };
-      if (query.rawSql && !query.builderOptions) {
-        return (
-          <div data-testid="query-editor-section-sql">
-            <SQLEditor {...props} />
-          </div>
-        );
-      }
-      if (!query.rawSql || !query.builderOptions) {
-        newQuery = {
-          ...newQuery,
-          rawSql: defaultCHBuilderQuery.rawSql,
-          builderOptions: {
-            ...defaultCHBuilderQuery.builderOptions,
-            database: props.datasource.getDefaultDatabase() || 'default',
-          },
-        };
-      }
-      return (
-        <div data-testid="query-editor-section-builder">
-          <QueryBuilder
-            datasource={props.datasource}
-            builderOptions={newQuery.builderOptions}
-            onBuilderOptionsChange={onBuilderOptionsChange}
-            format={newQuery.format}
-            app={app}
-          />
-          <Preview sql={newQuery.rawSql} />
-        </div>
-      );
+  /**
+   * Grafana will sometimes replace the builder options directly, so we need to sync in both directions.
+   * For example, selecting an entry from the query history will cause the local state to fall out of sync.
+   * The "key" property is present on these historical entries.
+   */
+  const queryKey = query.key || ''
+  const lastKey = useRef<string>(queryKey);
+  if (queryKey !== lastKey.current && query.editorType === EditorType.Builder) {
+    builderOptionsDispatch(setAllOptions((query as CHBuilderQuery).builderOptions || {}));
+    lastKey.current = queryKey;
   }
-};
 
-export const CHQueryEditor = (props: CHQueryEditorProps) => {
-  const { query, onChange, onRunQuery } = props;
+  // Prevent trying to run empty query on load
+  const shouldSkipChanges = useRef<boolean>(true);
+  if (isBuilderOptionsRunnable(builderOptions)) {
+    shouldSkipChanges.current = false;
+  }
 
-  React.useEffect(() => {
-    if (typeof query.selectedFormat === 'undefined' && query.queryType === QueryType.SQL) {
-      const selectedFormat = Format.AUTO;
-      const format = getFormat(query.rawSql, selectedFormat);
-      onChange({ ...query, selectedFormat, format });
+  useEffect(() => {
+    if (shouldSkipChanges.current || query.editorType === EditorType.SQL) {
+      return;
     }
-  }, [query, onChange]);
+
+    const sql = generateSql(builderOptions);
+    onChange({
+      ...query,
+      editorType: EditorType.Builder,
+      rawSql: sql,
+      builderOptions,
+      format: mapQueryTypeToGrafanaFormat(builderOptions.queryType)
+    });
+
+    // TODO: fix dependency warning with "useEffectEvent" once added to stable version of react
```

Comment:
Yes it is, but the `useEffectEvent` hook is still experimental and not available in our version:

https://react.dev/learn/separating-events-from-effects#declaring-an-effect-event

This would likely be the best solution to this portion of code, although I would probably need to re-visit it again to be sure. Right now the dependency array isn't being used correctly which can lead to unexpected bugs if someone tries to modify this in the future